### DISCUSSION TITLE
SML (Smart Meter) does not check checksum?!

### DISCUSSION THREAD

**Discussion by toastbrot69** [arendst/Tasmota#19321] [Ideas]
### PROBLEM DESCRIPTION
The SML protocol implementation (for smart meters) seems not to check the protocols checksum, leading to erroneous meter readings

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x ] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [ x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [ x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [ x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [ x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [ x] Device used (e.g., Sonoff Basic): esp-01
- [x ] Tasmota binary firmware version number used: 13.0.0
  - [ ] Pre-compiled
  - [x ] Self-compiled
- [ x] Flashing tools used: ota
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:

18:34:32.581 CMD: Backlog Template; Module; GPIO 255
18:34:32.609 MQT: stat/zaehlerEG/RESULT = {"NAME":"Generic","GPIO":[1,1,1,1,1,1,1,1,1,1,1,1,1,1],"FLAG":0,"BASE":18}
18:34:32.843 MQT: stat/zaehlerEG/RESULT = {"Module":{"1":"Sonoff Basic"}}
18:34:33.099 MQT: stat/zaehlerEG/RESULT = {"GPIO0":{"32":"Button1"},"GPIO1":{"0":"None"},"GPIO2":{"0":"None"},"GPIO3":{"0":"None"},"GPIO4":{"0":"None"},"GPIO5":{"0":"None"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO12":{"224":"Relay1"},"GPIO13":{"320":"Led_i1"},"GPIO14":{"0":"None"},"GPIO15":{"0":"None"},"GPIO16":{"0":"None"},"GPIO17":{"0":"None"}}



```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:


```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
just wait a while, place a piece of paper between the sensor and the meter to lower the S/N and increase the error rate

### EXPECTED BEHAVIOUR
A quick glance at the source shows, that the SML parser just scans for lines starting with 0x77, ignoring the SML-file structure?
The complete SML file should be parsed and the checksum (0x1a line after postfix) evaluated, and parse the 0x77 lines only if the checksum is correct…

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

18:37:08.452 MQT: tele/zaehlerEG/SENSOR = {"Time":"2023-08-15T18:37:08","DWS7412":{"energy":137.4220,"power":394.85,"meter_id":"445a47"}}
18:37:23.409 MQT: tele/zaehlerEG/SENSOR = {"Time":"2023-08-15T18:37:23","DWS7412":{"energy":137.4236,"power":395.33,"meter_id":"445a47"}}
18:37:38.421 MQT: tele/zaehlerEG/SENSOR = {"Time":"2023-08-15T18:37:38","DWS7412":{"energy":5.8245,"power":999.999,"meter_id":"005a47"}}      <—— wrong values, energy must increase, power is hilarious 
18:37:53.429 MQT: tele/zaehlerEG/SENSOR = {"Time":"2023-08-15T18:37:53","DWS7412":{"energy":137.4269,"power":392.06,"meter_id":"445a47"}}
18:38:08.441 MQT: tele/zaehlerEG/SENSOR = {"Time":"2023-08-15T18:38:08","DWS7412":{"energy":137.4285,"power":396.81,"meter_id":"445a47"}}

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

Thanks!

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by gemu2015** [comment]
see this issue
https://github.com/arendst/Tasmota/issues/16668

**Reply by toastbrot69** [reply]
Yes, that’s the point. A median filter TRIES to filter out irregular (wrong) values (only working if the filter is long enough and there are more correct than erroneous blocks in that range). But that’s fighting symptoms, not the problem.
One does not need to store the whole SML file for crc calculation or parsing: starting crc-calculation with the frame-start, store only the wanted values, and if then in the end the crc is correct, forward the values to mqtt/script…
A crc is exactly for that problem: distinguish corrupt from correct data without assumptions and statistics.

**Reply by gemu2015** [reply]
the problem is improper hardware. serial transfer is extremely robust if signals are well formed.
i never ever have wrong readings in my SML meter.  that's also true for the vast majority of users. try to ensure that your IR head is well designed and positioned and use hardware serial.
recoding the SML driver for that purpose is a bit of work.  i am not willing to invest that time.
feel free to code your own driver and make a pr.

**Comment by maikischa** [comment]
I have the same issue as I transmit my values vie LoRa and revive sometimes wrong packages. This issue goes so far that the esp8266 and esp32 crashes.
I‘m using now this Arduino sketch (https://github.com/aldadic/esp-smartmeter-reader) witch implements cry checks and word flawlessly.  By the way this is the dump after the crash:

Exception (3):
epc1=0x4027f6e9 epc2=0x00000000 epc3=0x00000000 excvaddr=0x40000000 depc=0x00000000

>>>stack>>>

ctx: cont
sp: 3ffff8e0 end: 3fffffc0 offset: 0190
3ffffa70:  [00000000](tel:00000000) a0000000 [00000000](tel:00000000) 0000001c  
3ffffa80:  f53ceb68 52293966 57550a7b 6f6c5548  
3ffffa90:  19e75704 ca9e9d02 f278897d c799c254  
3ffffaa0:  f6cd19c1 eba64980 03346d44 2b5c6456  
3ffffab0:  7cc45a34 78145000 e186fb28 9f73f52a  
3ffffac0:  dffcdd2a 93fd73a0 0434cea4 fc07766d  
3ffffad0:  a99557ed 6717cf1b f833b8c9 ce8298f6  
3ffffae0:  ada19949 9b10b976 36b1203f 25732c20  
3ffffaf0:  b66ee03f b7eaa995 d8f3e8e6 931dcc1f  
3ffffb00:  6f194173 929985b5 6e9d08d9 fc048d6c  
3ffffb10:  332f4eca 33200bee 71d0235f 68874cf8  
3ffffb20:  000f4524 19576fa7 42ff6d95 5ba74716  
3ffffb30:  19582a83 5372f4cc 77d004cc fac40b8e  
3ffffb40:  1f32e116 24a2f000 e5f6ea98 a9b6ff42  
3ffffb50:  68e2e5da c1541a98 ea79d6d0 7c776fc3  
3ffffb60:  275ec319 a35f969c 4d4ba40c 7a9763be  
3ffffb70:  89137d28 f73b2466 3fffc228 401063b1  
3ffffb80:  0000000e [00000000 00000010](tel:00000000 00000010) [00000000](tel:00000000)  
3ffffb90:  f73b2466 89137d28 7a9763be 4d4ba40c  
3ffffba0:  f2a7c231 c799c254 ffff7ae0 b61ec54a  
3ffffbb0:  9f73f52a 3ffffda0 40000004 0434cea4  
3ffffbc0:  fc07766d b66ee03f 25732c20 6717cf1b  
3ffffbd0:  a99557ed d8f3e8e6 b7eaa995 ada19949  
3ffffbe0:  9b10b976 929985b5 6e9d08d9 f833b8c9  
3ffffbf0:  ce8298f6 36b1203f 931dcc1f 6f194173  
3ffffc00:  fc048d6c 3ffffb78 13af618c 52293966  
3ffffc10:  03346d44 ea9f4759 97a20fd6 be3bb7e1  
3ffffc20:  3ffffc70 3ffffc74 3ffffc50 fffffff0  
3ffffc30:  3fff7af4 [00000000](tel:00000000) 3ffffd68 4027ef4a  
3ffffc40:  3ffffd74 [00000010](tel:00000010) 3ffffc70 fffffffc  
3ffffc50:  [00000000 00000000](tel:00000000 00000000) [00000000 00000000](tel:00000000 00000000)  
3ffffc60:  3fff374c 3fff7af4 3ffffe80 4025c8ef  
3ffffc70:  402ae284 7e19dc00 2de9beeb b2cbc247  
3ffffc80:  1daeafe2 9b6044a4 b689fa4f [04423808](tel:04423808)  
3ffffc90:  19ec97ea 57e8c370 e161393f e5230137  
3ffffca0:  fccf96dd d97802c0 38193bff dd3a3ac8  
3ffffcb0:  21f5ac15 37e95b3d 0ff060c2 d2ca5a0a  
3ffffcc0:  f33ff61f 52ab9b30 5d5bfbf2 8f91a1f8  
3ffffcd0:  7cae57e7 96f00f20 cbabf4d2 443a552a  
3ffffce0:  389402cd f487b227 3f2c46f5 7b1613df  
3ffffcf0:  43821112 67057b3d 58293dc8 233f2e17  
3ffffd00:  60bd3f05 067010ed 5e592d25 7d660332  
3ffffd10:  1ddb3c37 899b8a49 d7c2a76c aaa4a45e  
3ffffd20:  b77f9869 3ffe8340 [00000035](tel:00000035) 402734e5  
3ffffd30:  3ffffe60 3ffffe40 [00000004](tel:00000004) 3ffe8340  
3ffffd40:  3ffef650 [00000015 00000000](tel:00000015 00000000) 40251fc1  
3ffffd50:  [00000688](tel:00000688) 000000d1 000000d1 [00000016](tel:00000016)  
3ffffd60:  402b6135 0000000a 402ae260 3ffffc70  
3ffffd70:  4027f418 1bcf1767 ed5795a9 6d7607fc  
3ffffd80:  a4ce3404 694b5349 e0634975 62206425  
3ffffd90:  [00000002 00000000](tel:00000002 00000000) [00000000 40102858](tel:00000000 40102858)  
3ffffda0:  [00000000 00000000](tel:00000000 00000000) [00000000 00000000](tel:00000000 00000000)  
3ffffdb0:  [00000001 00000002](tel:00000001 00000002) [00000000 00000000](tel:00000000 00000000)  
3ffffdc0:  [00000000 00000000](tel:00000000 00000000) 3ffedf20 [00040000](tel:00040000)  
3ffffdd0:  4026ae1a [00000001](tel:00000001) 3fff520c 4026ae10  
3ffffde0:  [00000688](tel:00000688) 4b534924 63497569 206425e0  
3ffffdf0:  4026ae62 [00000001](tel:00000001) 3fff520c 4026ae10  
3ffffe00:  3fff7ae4 [00000000 00000010](tel:00000000 00000010) [00000000](tel:00000000)  
3ffffe10:  [00000500](tel:00000500) 3fff7fc4 000004ff fffffffe  
3ffffe20:  [00000001](tel:00000001) 3fff7fc4 3fff789c 3ffffe80  
3ffffe30:  3fff7ae4 3fff7fc4 000000db 4025bfa0  
3ffffe40:  3fff7ac4 [00000000 00000500](tel:00000000 00000500) 4025ce90  
3ffffe50:  ffffffff 3fff7fc4 3fff7ac4 40204e69  
3ffffe60:  [00000001](tel:00000001) 3fff7fc4 3fff7ac4 fffffffe  
3ffffe70:  [00000001](tel:00000001) 3fff7fc4 3fff7ac4 4025c300  
3ffffe80:  000c0000 [00000000](tel:00000000) 694b5349 e0634975  
3ffffe90:  3ffffec0 3ffffec4 0000000b 3fff7ae4  
3ffffea0:  [00000000](tel:00000000) 3fff0bd4 3fff2738 ffffffdf  
3ffffeb0:  3ffef7a8 [00000000](tel:00000000) 3ffef860 402370c9  
3ffffec0:  4026ad10 [00000000](tel:00000000) 4bc6a7f0 [00000000](tel:00000000)  
3ffffed0:  [00000016 00000000](tel:00000016 00000000) 3ffef860 4026641b  
3ffffee0:  [00000000 00000006](tel:00000000 00000006) 3ffef860 ffffffdf  
3ffffef0:  3ffef7a8 [00000000](tel:00000000) 3ffef860 402374f1  
3fffff00:  40106271 000e0561 4010136e 40231dab  
3fffff10:  [00000000](tel:00000000) 4bc6a7f0 00000b9d 00000006  
3fffff20:  [00000016](tel:00000016) 3ffef7a8 [00000000](tel:00000000) 4024f54f  
3fffff30:  [00000016 00000006](tel:00000016 00000006) [00000000](tel:00000000) 4022fd9d  
3fffff40:  [00000000 00000000](tel:00000000 00000000) 4bc6a7f0 3fff2738  
3fffff50:  3fffdad0 [00000000 00000006](tel:00000000 00000006) 40231fc8  
3fffff60:  [00000000](tel:00000000) 1264780e 00000b9d 4025259a  
3fffff70:  3fffdad0 [00000000](tel:00000000) 00000b9d 40252650  
3fffff80:  [00000000 00000000](tel:00000000 00000000) [00000001](tel:00000001) 40100be4  
3fffff90:  3fffdad0 [00000000](tel:00000000) 3fff26f8 3fff2738  
3fffffa0:  3fffdad0 [00000000](tel:00000000) 3fff26f8 40268d20  
3fffffb0:  feefeffe feefeffe 3ffe8728 40101ffd  
<<<stack<<<

--------------- CUT HERE FOR EXCEPTION DECODER ---------------
H!)Q���&

00:00:00.001 HDW: ESP8266EX
00:00:00.013 UFS: FlashFS mounted with 1984 kB free
00:00:00.075 CFG: Loaded from File, Count 176

**Comment by tombueng** [comment]
You can check out my PR that adds CRC checks to Tasmota SML here 
https://github.com/arendst/Tasmota/pull/23205
